### DISCUSSION TITLE
Azure IPAM solution/use case context

### DISCUSSION THREAD

**Discussion by brianmooremsft** [Azure/ipam#11] [General]
Apologies if I missed it in docs somewhere, but was wondering how the architects of this solution see it fitting into the larger world of IPAM solutions that enterprises customers already use to manage their traditional on-prem networks (ie Infoblox).  Is the use case  this Azure IPAM service/solution just for customers who may not have an existing enterprise class solution, or is it thought that this Azure IPAM solution would manage just the customer's Azure IP space and provide some "integration" to upstream IP address managers?

**Comment by DCMattyG** [comment]
Hey @brianmooremsft, I believe the answer is 'yes' to both. There's obviously a number of enterprise solutions out there that work well with Azure (Infoblox, etc.), but a few use cases here would be:

1. Customer uses enterprise appliance for on-premises, but hasn't/won't extend that into Azure.
2. Customer is cloud-native (Azure) and wants visibility into all of the vNets/Subnets/IP Endpoints within their cloud space.

A discussion point internally, and with a few potential customers is indeed "I'd like to export the data from *this* solution into my <fill in the blank> enterprise appliance". So that would be something we can explore through API integrations in the future.

I would LOVE any additional thoughts and/or feedback you may have here. We definitely feel that having a holistic view of the IP Space in Azure is a current pain point, especially when you have various teams spinning up subscriptions and creating vNet's, whom might not have knowledge of the overall IP landscape for their organization, and the implications of not selecting non-overlapping IP space as they build out their solutions (e.g. problems peering vNets, inability to connect Azure to on-prem via VPN/ExpressRoute).

**Comment by brianmooremsft** [comment]
Interesting.  Let me connect with you internally, as we have current engagement w/ customer who is struggling with this and their current IPAM solution, and see if I can define some use cases that make sense.

I worked on the customer side for years with Infoblox and Azure, and never saw a good story there.

**Comment by dgitw** [comment]
Great to see MS picking this up and driving it.

Is there anything in the roadmap to automate IP request / allocation particularly in relation to Enterprise Landing Zone concept, similar to the solution @stefanstranger has come up with https://github.com/stefanstranger/AIPAS, he has nailed our use case

**Reply by DCMattyG** [reply]
Yes @dgitw. I actually looked at @stefanstranger's solution as we begin the work on this IPAM solution. The goal here is to create more of a "traditional" IPAM solution, complete with a UI, and higher level concepts like assigned IP Blocks, and even logical constructs where there "should" be non-overlapping IP space (akin to what Infoblox has with Address Spaces in BloxOne). An integration has been planned with ESLZ by adding some steps within the Landing Zone template to "call out" to Azure IPAM and retrieve the next available CIDR block within a given IP Block or a provided size, and use that to deploy the Landing Zone template. Any and all feedback is welcome, but that is what the current plan is for now.

**Reply by stefanstranger** [reply]
I'll try to test this solution too and provide feedback when comparing what I've developed in my [https://github.com/stefanstranger/AIPAS](https://github.com/stefanstranger/AIPAS) solution.

**Reply by DCMattyG** [reply]
That's very awesome of you @stefanstranger. Please reach out to myself or @hbendana if you run into any issues!

**Reply by stefanstranger** [reply]
Hi,

I run the deploy.ps1 file and got the following errors.

![image](https://user-images.githubusercontent.com/2125385/175261435-75823b86-c635-4c74-9289-9486931a27d7.png)

**Reply by DCMattyG** [reply]
Hi @stefanstranger, just to double check you've verified all of the prerequisites, correct?

https://azure.github.io/ipam/#/deployment/README?id=prerequisites

**Reply by hbendana** [reply]
Hi @stefanstranger, could you provide information regarding versions of the following please?
- PowerShell
- PowerShell Az Module
- PowerShell Microsoft.Graph Module
Also, can you share with us the type of Azure Subscription you are attempting to deploy the solution to? (for example...Commercial, MSDN, MSFT internal) I suspect you are running into either a module version issue or a permissions issue regarding Microsoft Graph and the ability to grant consent to the app registrations.

Thank you!

**Comment by PStenbumling** [comment]
Any update on this ?

**Reply by DCMattyG** [reply]
Hi @P-Stenbumling, we're just finalizing the documentation and should have the code pushed to our main branch and ready for testing by the end of this week/Monday at the latest. I'll drop a note in this thread as soon as the code is live. Thanks for checking in, excited to get this finally launched!

**Comment by DCMattyG** [comment]
@brianmooremsft, @dgitw, and @P-Stenbumling, the initial code push is complete. Please take a look at our docs which can be found here: https://azure.github.io/ipam/

Looking forward to your feedback!

**Reply by vpgohil** [reply]
@DCMattyG , I tried latest code in our env and I see all vnets and subnets are loaded from azure, only issue I see is with scroll if I scroll down I only see limited Vnets and subnets.
I can other data if I sort them by name or use filter.

Also is there any way to exclude some of the subscriptions by adding some parameters ?

**Reply by DCMattyG** [reply]
Hi @vpgohil, I just wanted to clarify a few items from your note above?

**"...I only see limited Vnets and subnets."**
Here are you saying that in the IPAM interface you are seeing less vNets/Subnets than expected? If so, is what you see in IPAM equivalent, less, or more that you're able to see if you login to port.azure.com?

**"...is there any way to exclude some of the subscriptions...?"**
Are you asking to be able to filter the vNets section by subscription ID? Or are you asking to exclude some subscriptions from the IPAM view? If you're asking to exclude some subscriptions from the IPAM view, is this something you'd like to set per user, or globally for all IPAM users?

Thanks you so much for the feedback, and we're looking forward to making this an amazing experience for all of you!

**Reply by vpgohil** [reply]
@DCMattyG ,

I see less vNets and Subnets in IPAM.

I want to exclude some dev/test subscription globally from IPAM.

also another issue I see is with adding admin user search return only users whose name starts with A (it could also mean it is returning limited number of users).

**Reply by DCMattyG** [reply]
Thanks for that info @vpgohil!

The less vNets/Subnets is likely due to paging that is happening for the Azure Resource Graph results. I'm working on a fix for that now.

The limited users in the Admin tab is likely due to the limited results the Microsoft Graph API returns at once. I'm working on a fix for that now.

I'm thinking through the best way to implement a way to exclude some subscriptions. When you are logged in as an admin, you are leveraging the Service Principal credentials which is sitting at the root Management Group level, so it has visibility to everything. When you are logged in as a normal user, you only see what your authenticated user would normally see in the Azure portal. Just some information for you in the meantime while I think through a solution here.

**Reply by vpgohil** [reply]
@DCMattyG , thanks for quick response!

I see reservation through API is working fine but I cannot find way to reserve from UI.

**Reply by DCMattyG** [reply]
At this time, we haven't added that functionality to the UI. The reason behind that is, we had imagined reservations happening almost exclusively as part of some form of automation, whether that be embedded in an ARM/Bicep/Terraform template to deploy (for example) Landing Zones, and perhaps via Azure Automation/Functions to automate vNet creation.

That said, if you have a use case in mind where it would make sense for users to come to the UI to create reservations, there is no opposition to adding that capability.
